notifications_utils/recipients.py

```diff
@@ -225,71 +202,38 @@ def more_rows_than_can_send(self):
     def too_many_rows(self):
         return len(self) > self.max_rows
 
-    def get_annotated_rows(self):
-        if self.too_many_rows:
-            return []
-        for row_index, row in enumerate(self.rows):
-            if self.template:
-                self.template.values = dict(row.items())
-            yield dict(
-                columns=Columns({key: {
-                    'data': value,
-                    'error': self._get_error_for_field(key, value),
-                    'ignore': (
-                        key not in self.placeholders_as_column_keys
-                    )
-                } for key, value in row.items()}),
-                index=row_index,
-                message_too_long=bool(
-                    self.template and
-                    self.template.is_message_too_long()
-                )
-            )
-
     @property
-    def initial_annotated_rows(self):
-        for row in self.annotated_rows:
-            if row['index'] < self.max_initial_rows_shown:
-                yield row
+    def initial_rows(self):
+        return islice(self.rows, self.max_initial_rows_shown)
 
     @property
-    def annotated_rows_with_errors(self):
-        for row in self.annotated_rows:
-            if RecipientCSV.row_has_error(row):
-                yield row
+    def displayed_rows(self):
+        if any(self.rows_with_errors) and not self.missing_column_headers:
+            return self.initial_rows_with_errors
+        return self.initial_rows
+
+    def _filter_rows(self, attr):
+        return filter(lambda row: getattr(row, attr), self.rows)
```

Comment:
but [guido doesn't like lambdas](https://www.artima.com/weblogs/viewpost.jsp?thread=98196) :(

`return (row for row in self.rows if getattr(row, attr))`

(i don't mind if you keep the filters tho)

Reply:
Changed in 8fb32c0fad99a68b9750db04a30de8c4bb1bf12b